students/tri_nguyen/lesson04/mailroom.py

```diff
@@ -0,0 +1,175 @@
+    4. {}
+    '''.format(*ACTIONS))
+
+
+def display_donor_list():
+    ''' this function displays list of donors
+        from donor_dict '''
+
+    for idx, item in enumerate(donor_dict):
+        print(idx + 1, '-', donor_dict[idx]['name'])
+
+
+def send_thanks():
+    ''' this function composes an email thanking
+        the donor for their generous donation '''
+
+    name_list = [item['name'] for item in donor_dict]
+
+    while True:
+
+        full_name = input('Enter a Full Name (or type "list" to get the list of donors): ').title()
+
+        if full_name.lower() == 'list':
+            display_donor_list()
+            continue
+
+        elif full_name not in name_list:
+            donation_amount = float(input('Enter donation amount: '))
+            donor_dict.append(dict(name=full_name.title(), donation=donation_amount, count=1, average=donation_amount))
+
+        elif full_name in name_list:
+            donation_amount = float(input('Enter donation amount: '))
+            for item in donor_dict:
+                if item['name'] == full_name:
+                    item['donation'] = item['donation'] + donation_amount
+                    item['count'] = item['count'] + 1
+                    item['average'] = item['donation'] / item['count']
+
+        choice = input('Do you want to quit and send email?(y/n) ')
+
+        if choice == 'y':
+            email_template = '''
+                Dear {0},
+
+                Thank you for giving us ${1:,.2f}. Your money will be put to good use.
+                Please come back and donate more.
+
+                Best Regards,
+
+                John Doe
+            '''
+            print(email_template.format(full_name, float(donation_amount)))
+
+            break
+        else:
+            continue
+
+
+def create_report():
+    ''' this function generate a report of donors '''
+
+    template = '{0:<26}|{1:^13}|{2:^11}|{3:^14}'
+    value_template = '{0:<26} ${1:12.2f} {2:11} ${3:13.2f}'
+
+    formatted_columns = template.format('Donor Name', 'Total Given', 'Num Gifts', 'Average Gift')
+
+    print(formatted_columns)
+
+    print('{}'.format('-' * len(formatted_columns)))
+
+    print('Report of Donors is below:\n')
+
+    donor_list = [list(item.values()) for item in donor_dict]
+
+    donor_list.sort(key=itemgetter(1), reverse=True)
+
+    for item in donor_list:
+        print(value_template.format(item[0], item[1], item[2], item[3]))
+    print()
+
+
+def send_all_letters():
+    ''' this function send letters to everyone '''
+
+    email_dict = {
+        'greetings': 'Dear',
+        'body_one': 'Thank you for your kind donation of ',
+        'body_two': 'It will be put to very good use.',
+        'end': 'Sincerely',
+        'signature': '-The Team'}
+
+    value_list = [list(item.values()) for item in donor_dict]
+
+    for each_list in value_list:
+
+        each_list[0] = each_list[0].split()
+
+        template = '{}_' * (len(each_list[0]) - 1) + '{}.txt'
```

Comment:
for your question on `,` in the file name:
```
>>> s = 'William_Gates,_III.txt'
>>> s.replace(',', '')
'William_Gates_III.txt'
```
you can use replace to just remove the comma, there are also other ways to do it if you have a set of chars you want to remove.